llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp

```diff
@@ -84,6 +86,42 @@ ModulePass *llvm::createSYCLLowerWGLocalMemoryLegacyPass() {
   return new SYCLLowerWGLocalMemoryLegacy();
 }
 
+// In sycl header __sycl_allocateLocalMemory builtin call is wrapped in
+// group_local_memory/group_local_memory_for_overwrite functions, which must be
+// inlined first before each __sycl_allocateLocalMemory call can be lowered to a
+// unique global variable. Inlining them here so that this pass doesn't have
+// implicit dependency on AlwaysInlinerPass.
+static bool inlineGroupLocalMemoryFunc(Module &M) {
+  Function *ALMFunc = M.getFunction(SYCL_ALLOCLOCALMEM_CALL);
+  if (!ALMFunc || ALMFunc->use_empty())
+    return false;
+
+  bool Changed = false;
+  for (auto *U : ALMFunc->users()) {
+    auto *Caller = cast<CallInst>(U)->getFunction();
+    if (!Caller->hasFnAttribute(Attribute::AlwaysInline)) {
+      // Already inlined.
+      continue;
+    }
+    std::string FName = llvm::demangle(Caller->getName());
+    if (FName.find("sycl::_V1::ext::oneapi::group_local_memory") ==
```

Comment:
Hardcoding current function name from DPC++ library is unfortunate. The code in the DPC++ header files can be changed at any time.

To make it more robust, I thought we could go up in the call stack up-to the kernel function ignoring all functions in `sycl::` namespace. This will require SYCL kernel to be inlined into kernel function wrapper.

@Naghasan, do you have any thoughts on that?

Reply:
I agree it is unfortunate, especially w.r.t. upstreaming. I don't know what the plans are for this one but if it is seen as important, we might want to improve this.

> This will require SYCL kernel to be inlined into kernel function wrapper.

I don't think this is an issue TBH, I don't see any benefit in not inline the SYCL kernel in the wrapper, even in SPIR-V.

I think relying on an attribute is probably the most flexible: this makes the compiler agnostic to header refactor and changes in API. It is also cheap to add.

I also just realized `syclcompat::local_mem` uses it, it isn't technically a valid usage of it w.r.t. the extension but something the attribute would allow to correctly handle.

cc @elizabethandrews @joeatodd

Reply:
> I think relying on an attribute is probably the most flexible: this makes the compiler agnostic to header refactor and changes in API. It is also cheap to add.

A new attribute "sycl_forceinline" is added in https://github.com/intel/llvm/pull/16356/commits/a4fe9152efda3e8a28d78b736a4b51d02be82157
Please review.